tests/python/test_fields_builder.py

```diff
@@ -3,174 +3,259 @@
+    assign_mutliple_field_kernel_func1(5, 5)
+    for i in range(shape_size_1d):
         assert y[i] == i + 10
         assert z[i] == i + 5
 
-    func1()
-    for i in range(n):
+    assign_single_field_kernel_func()
+    for i in range(shape_size_1d):
         assert x[i] == i * 3
 
-
+'''
+Test fields with builder destory.
+'''
 @ti.test(arch=[ti.cpu, ti.cuda, ti.vulkan])
 def test_fields_builder_destroy():
-    def A(i):
-        n = i * 10**3
+
+    # [destroy] Args of main body
+    test_sizes = [1] # [1, 2, 3]
+    size_extend_factor = 1 # 10 ** 3
+    # note: currently only consider preicison all paltform supported,
+    # more detailed here: https://docs.taichi.graphics/lang/articles/basic/type#supported-primitive-types
+    field_types = [ti.f32, ti.i32]
+    field_sizes = [1] # [1, 5, 10]
+
+   # [destroy] 1. test for single destroy multiple fields
+    def test_for_single_destroy_multi_fields(test_sizes, size_extend_factor, field_types, field_sizes):
         fb = ti.FieldsBuilder()
-        a = ti.field(ti.f64)
-        fb.dense(ti.i, n).place(a)
+        for tsize_idx in range(len(test_sizes)):
+            for ftype_idx in range(len(field_types)):
+                for fsize_idx in range(len(field_sizes)):
+                    # init
+                    test_1d_size = test_sizes[tsize_idx] * size_extend_factor
+                    field_type = field_types[ftype_idx]
+                    field_size = field_sizes[fsize_idx]
+
+                    for create_field_idx in range(field_size):
+                        field = ti.field(field_type)
+                        fb.dense(ti.i, test_1d_size).place(field)
+                    fb_snode_tree = fb.finalize()
+        fb_snode_tree.destroy()
+
+    test_for_single_destroy_multi_fields(test_sizes, size_extend_factor, field_types, field_sizes)
+
+    # [destroy] 2. test for multiple destroy for multiple fields
+    def test_for_multi_destroy_multi_fields(test_sizes, size_extend_factor, field_types, field_sizes): #size_1d_0, size_1d_1):
+        fb0 = ti.FieldsBuilder()
+        fb1 = ti.FieldsBuilder()
+
+        for tsize_idx in range(len(test_sizes)):
+            for ftype_idx in range(len(field_types)):
+                for fsize_idx in range(len(field_sizes)):
+                    test_1d_size = test_sizes[tsize_idx] * size_extend_factor
+                    field_type = field_types[ftype_idx]
+                    field_size = field_sizes[fsize_idx]
+
+                    for create_field_idx in range(field_size):
+                        field0 = ti.field(field_type)
+                        field1 = ti.field(field_type)
+
+                        fb0.dense(ti.i, test_1d_size).place(field0)
+                        fb1.pointer(ti.i, test_1d_size).place(field1)
+
+                        fb0_snode_tree = fb0.finalize()
+                        fb1_snode_tree = fb1.finalize()
+
+        # destroy
+        fb0_snode_tree.destroy()
+        fb1_snode_tree.destroy()
```

Comment:
Is here a proper destory place?

Reply:
It is OK to destroy here. However, you cannot call `.dense()` or `.pointer()` after calling `.finalize()`. This will trigger `InvalidOperationError`.

Reply:
Ok